src/gridfs/download.rs

```diff
@@ -139,16 +142,214 @@ impl GridFsBucket {
+
+enum State {
+    // Idle stores an Option<Cursor> so that the cursor can be moved into a GetBytesFuture without
+    // requiring ownership of the state. It will always store a value when the stream's state is
+    // Idle and can be unwrapped safely.
+    Idle(Box<Option<Cursor<Chunk<'static>>>>),
+    Busy(GetBytesFuture),
+    Done,
+}
+
+impl State {
+    fn set_busy(&mut self, new_future: GetBytesFuture) -> &mut GetBytesFuture {
+        *self = State::Busy(new_future);
+        match self {
+            Self::Busy(ref mut future) => future,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl GridFsDownloadStream {
+    async fn new(
+        file: FilesCollectionDocument,
+        chunks: &Collection<Chunk<'static>>,
+    ) -> Result<Self> {
+        let initial_state = if file.length == 0 {
+            State::Done
+        } else {
+            let options = FindOptions::builder().sort(doc! { "n": 1 }).build();
+            let cursor = chunks.find(doc! { "files_id": &file.id }, options).await?;
+            State::Idle(Box::new(Some(cursor)))
+        };
+        Ok(Self {
+            state: initial_state,
+            cached_bytes: None,
+            current_n: 0,
+            file,
+        })
+    }
+
+    /// Gets the file `id` for the stream.
+    pub fn files_id(&self) -> &Bson {
+        &self.file.id
+    }
+}
+
+impl AsyncRead for GridFsDownloadStream {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::result::Result<usize, futures_util::io::Error>> {
+        let stream = self.get_mut();
+
+        if let Some(cache) = &mut stream.cached_bytes {
+            let bytes_to_write = std::cmp::min(cache.len(), buf.len());
+            buf[..bytes_to_write].copy_from_slice(cache.drain(0..bytes_to_write).as_slice());
+            if cache.is_empty() {
+                stream.cached_bytes = None;
+            }
+            return Poll::Ready(Ok(bytes_to_write));
+        }
+
+        let future = match &mut stream.state {
+            State::Idle(cursor) => {
+                let chunks_in_buf =
+                    FilesCollectionDocument::n_from_vals(buf.len() as u64, stream.file.chunk_size);
+                // We should read from current_n to chunks_in_buf + current_n, or, if that would
+                // exceed the total number of chunks in the file, to the last chunk in the file.
+                let final_n = std::cmp::min(chunks_in_buf + stream.current_n, stream.file.n());
+                let n_range = stream.current_n..final_n;
+                let total_chunks = n_range.end - n_range.start;
+
+                let new_future = get_bytes(
+                    cursor.take().unwrap(),
+                    n_range,
+                    stream.file.chunk_size,
+                    stream.file.length,
+                )
+                .boxed();
+
+                stream.current_n += total_chunks;
```

Comment:
I think this logic might be a little clearer if we just assign `current_n` to `final_n` directly.